README.md
# 文明6 Plus：和而不同 v1.0

### 简介
【文明6 Plus: 和而不同】是由一群有经验的PVE玩家针对原版文明6底层机制的诸多痛点而设计的一套更贴合人类发展实际、游戏各个元素更平衡、玩法更多元多样的大型模组。该模组包含的游戏规则解决了官方版本中较为突出的砍树、劫掠等问题，进行了大量平衡性的调整，并同时引入包容性的差异化发展作为特色。“始于删砍树，却绝不止于删砍树“，模组在减少无聊而困难的重复回合操作、让玩家把精力真正的集中在规划与抉择中的同时，力求做到在游戏进程的各个阶段提供给玩家因地制宜的选择空间，各种发展模式都独具特色而又不可替代，期望能让玩家更好地、真正地体会到《文明》的游戏精髓与乐趣。

### 在线百科
[请在该共享文档查看](https://docs.qq.com/sheet/DTGRlWFlsRHROTldV?tab=BB08J2)

### 模组特色
- 砍树砍石头等移除地貌和资源收获和劫掠的收益不再随游戏进度而变化：砍树不涨价，老马环游、木制火箭、刮地飞天成为过去式。
- 区域价格与游戏进度脱钩，与建造数量相关且价格更低：更符合实际的同时鼓励均衡发展。
- 平衡建筑收益，提供高级建筑辐射加成和百分比加成：大城市带动小城市、打造属于你的帝国心脏。
- 更多的改良区域相互配合，更合理的改良成长加成：由久经考验的文明6PVE骨灰级玩家打磨，给你带来规划上的高级体验。
- 重新平衡的宗教体系：各种万神因地制宜，各个信条各具特色。
- 地图生成、地形产出与出生地平衡：开图就能打，就算上沙下土也不用被迫刷图，就算是盘古也能铺开城市不跳水。
- 通过城市政策来对每个城市单独进行管理：每座城市都能发展，每座城市都有自己的特点。
- 重做的强力总督：每个总督都自成体系，赋予玩家更多的打法多样性。
- 科技人文树更新：科技人文树分叉更明显，发展combo差异化，同时突出更多的关键科技、人文节点，收益提升，让玩家不再为了科技而发展科技。
- 军事体系大改：告别单一小马征服，多兵种相互配合。
- 平衡的城邦、伟人收益：弱势城邦、伟人能力增强，旧的伟人，新的体验。
- 平滑宜居度分级，提升宜居度收益，让这一游戏元素重新焕发生机。
- 还有更多新奇观、新建筑等你探索。

### 获取方式
* Steam: 创意工坊搜索Harmony找到对应项目（注意要找1.1版本），进入项目页面后，点击订阅即自动安装。
* Epic: 请访问[在线百科的FAQ一栏](https://docs.qq.com/sheet/DTGRlWFlsRHROTldV?tab=aqb7re)

### 注意事项
* 至少需要拥有风云变幻和迭起兴衰DLC，我们的内容全部基于风云变幻规则集。
* 我们也支持了缺少其他DLC的情况，但我们制作时是基于全DLC的情况的，所以缺少其他DLC将可能导致体验上有差异。

### 联系我们
【和而不同】交流QQ群：1142970145

### 模组兼容性
我们将确认兼容的模组汇总到了下述两个合集内（合集均可一键订阅），对于合集以外的mod的兼容性无法保证。

#### 游戏性mod
* 合集：[支持的游戏性mod](https://steamcommunity.com/sharedfiles/filedetails/?id=2399536104)
* 我们对steam工坊中的一部分优秀的游戏性mod，如各种奇观、特色单位等进行了特殊支持，来作为内容的扩充并将其融入了和而不同的内容中。
* 关于具体的内容可查看[在线百科中的兼容模组部分](https://docs.qq.com/sheet/DTGRlWFlsRHROTldV?tab=8xb20k)。

#### 界面mod
* 合集：[实用界面mod合集](https://steamcommunity.com/sharedfiles/filedetails/?id=2268223724)

### 已整合第三方mod
* [城市道路(City Roads)](https://steamcommunity.com/sharedfiles/filedetails/?id=1284907124) (增加了轮子科技的要求)
* [黄金资源(Gold Resource)](https://steamcommunity.com/sharedfiles/filedetails/?id=870865055)
* [伙友骑兵bug修复(Hetairoi's general bug fix mod)](https://steamcommunity.com/sharedfiles/filedetails/?id=2183461036)
* [Expanded Initial Vision](https://steamcommunity.com/sharedfiles/filedetails/?id=970242434) （改为揭示3环视野，玛雅6环）

### 开发者列表
* DeepLogic
* 小小雨10101
* 六亿神州尽舜尧
* 傲娇的红莉栖
* 利牙菠萝
* 陈又
* wltk
* 小灰灰
* YT
* 柚子
* 奥伊米亚（配音）
* 黑毛毛小魔仙
* o(╯□╰)o
* Winged Hussar
* 榖音
* 法拉墨
* lp
* Vearmus
* Siv
* 流浪de蒲公英

### 致谢
感谢 Civ6ElitePlayers 群内的成员对模组的测试。

感谢 Hemmelfort 在 BiliBili 发布的MOD制作系列教程视频。

感谢所有在上述的游戏性mod合集和界面mod合集中的mod。

感谢下列MOD在包括内容实现方式等某些方面提供的借鉴（按字母序排列）。
* [黄金时代-美丽新世界](https://steamcommunity.com/sharedfiles/filedetails/?id=2209309479)
* AI+
* Botanist
* Better Balanced Game
* Better Balanced Start
* Better Trade Screen
* Cheat Menu Panel
* Chinese Translation of Wonder MODs
* Civilizations Expanded
* Customization VI
* Expanded Religions
* Extended TechTree
* Gift it to me
* Monopoly++: Resource Requirement Setup
* Real Strategy AI
* Real Tech Tree
* Religion Expansion


changelog.txt
Civilization VI game balancing mod (DL version)
Based on Civ6, Dec 2020 update (New Frontier Pass)

[Version v0.1.0]

[Gameplay]
Remove game cost escalation (except purchase plot).
移除买地消耗外的游戏进度的影响。

fixed cost of districts.
区域调整为固定价格

All districts that was having variant cost now have fixed cost 60 except
所有以前浮动价格的区域调整为60锤，水渠80，运河和大坝200。
aqueduct 80
canal 200
dam 200.

All projects cost 75 production except
所有项目调整为75锤，腓尼基迁都200锤，寻欢作乐美第奇宫廷盛会100锤。
move capital 200 (...)
court festival 100 (...)

If appliable, finish the project grants 
如果适用，完成项目给20点对应伟人点（若单一伟人）或10点对应伟人点（若多类伟人）
20 great person point (with single class)
10 great person point (with multiple class)

Pilaging gains stay fixed. (base 25)
劫掠获得固定收益，基础25。
change the raid & total war policy cards to 100% and 200% bonus on pillage gains.
更改劫掠卡（军事训练解锁）为100%额外劫掠收益，全面战争为200%额外。

[Yields]
change terrain & feature yields.
调整了地块和地貌产出
change citizen (specialist) yields, add 2 great person point yields for each type of specialist.
调整了专家产出，增加对应类型的2点伟人点产出。

[GoodyHuts]
Relic, builder, governor title may only be rewarded after 30, 15, 50 turns. (was 0, 0, 30)
30t, 15t, 50t 后才有可能从蘑菇中获得遗物、工人、总督头衔。
favor can be rewarded after 15 turns. (was 30)
外交点在 15t 后有可能从蘑菇中获得。

[Version v0.1.1]

Resume the free amenity for each city
恢复每个城市的一点免费宜居度，即前2人口不耗宜居度

Add worker's & military engineer's basic building charges
增加工人和军事工程师的一次使用次数

Increase the resource harvest & feature remove gains (from base 20 to 40 & to 30).
增加移除资源和地貌的收益（资源为基础40，地貌为基础30）

Update the bank cost from 290 to 250 (grand bazaar from 220 to 200), update the yield from 5 gold to 8 gold.
银行价格290锤调整为250锤（奥斯曼ub大巴扎由220锤调整为200锤），产出由5金调整为8金。
Update the workshop cost from 195 to 175.
工坊价格由195调整为175锤。

Fix the missing adjustment of Oppidum (Gaul). Adjust the cost of neighborhood and mbanza to cost 150 production.
修复遗漏的奥皮杜姆（高卢ud）的调整。社区和姆班赞（刚果ud）调整为150锤。

Increase the amenity of the (water) Entertainment Complex and the special ones. 
增加了（水上）娱乐中心的宜居度为3点，其对应ud给4点。

[Version v0.1.2]

Enable the amenity of arena (as well as tlachtli) being spread to range up to 4 tiles.
让竞技场（和蹴球场）的宜居度加成辐射4环。

Fix the BUG of great scientist IBN_KHALDUN.
修复了文艺复兴大科 伊本-赫勒敦 的效果。

Change the district cost of some districts to progression with extra 5 costs each copy, starting from 60 (60, 65, 70, ...),
for holy site, campus, encampment, commercial hub, entertainment complex, theater, industrial zone, water entertainment complex, and their corresponding unique districts.
将部分区域锤调整为等差函数，从60锤起，每多建造一个同种区域多5锤（即首个60锤，第二个65锤，第三个70锤，...），包括
圣地，学院，军营，商业，娱乐，剧院，工业，水上娱乐，和其对应的特色区域。

Introduce a new level of happiness: delighted, from amenity 1 to 2, while 0 amenity stays as content.
The delighted status gives 3% more non-food yields, and 10% growth rate, and 3 loyalty points each turn.
The happy & ecstatic status gives 20% & 30% growth rate, and 6 & 9 loyalty points each turn.
引入了新的一个宜居度等级：愉悦，宜居度为1-2时处于该状态，给非粮食产出+3%，给余粮+10%，每回合3点忠诚度。
高兴和欣喜若狂增加20%和30%的余量（城市增长速度），每回合6点和9点忠诚度。

[Version v0.1.3]
Balance update on units（单位平衡调整）：

units		cost 		old_cost	单位
setter		70 + 30n	80 + 30n	移民
worker  	45 + 3n		50 + 4n		工人
scout		25			30			侦察兵
slinger		30			35			投石手
babylonian	30			35			主力军团
warrior		35			40			勇士
trader		50 + 5n 	40 + scale	商人
spy			150 + 50n	225 + 75n	间谍

Balance update on buildings (建筑平衡调整)：
buildings	cost 		old_cost	建筑
MONUMENT	50			60			纪念碑
GRANARY		50			65			粮仓
WATER_MILL	70			80			水磨
PALGUM		70			80			沟渠
Give water mill one extra housing, one more production and one less food.
水磨+1住房，改为+2锤，自身不加粮食。

[Version v0.1.3.1]

disable the v0.1.3 changes. Introduce city policies (seperate mod)
取消 v0.1.3的改动，引入城市政策（单独mod）
Change the district cost of some districts to progression with extra 10 costs each copy. (was 5 cost)
部分区域每多建造一个同种区域多10锤（即首个60锤，第二个70锤，第三个80锤，...）（之前多5锤）

[Version v0.1.4]
fix the bug where the citizen inside unique district does not provide corresponding great person poinst.
修复了特色区域内的专家没有给对应伟人点产出的bug。

Give the district itself a citizen slot, and decrease the corresponding great person point yield of the district by one.
每个区域本体提供一个专家槽，并不再产出伟人点（除了俄罗斯的拉夫拉只减少了一点预言家点数）。

Adjust the specialist yields with new building added. The specialist yields about 2,3,4,6 values when the district, level one building, level two building, level three building have been built. The yield adjust slightly according the district type. The specialist require maintaince fee related to its yield.
调整了完成建筑对专家产出的影响。在拥有本体、一级、二级、三级建筑的情况下，专家产出大致为2、3、4、6点，不同类型的区域有调整。专家需要维护费，且随其产出增加而增加。

[Version v0.1.4.1]
Decrease the Trader starting cost to 40 production instead 50 production.
将商路基础锤减为40锤而不是50锤。

[Version v0.1.4.2]
Make the unique districts cheaper (starting cost from 60 -> 50, bath & mbanza 80% cheaper than its replacement district).
让特色区域（ud）更便宜，起始花费调整为50锤（减少10锤），浴场和姆班赞调整为80%的正常需求（浴场80锤，姆班赞120锤）。

Grant one trade route capacity to each commercial hub and harbor district (including the corresponding unique districts).
港口和商业区增加一条商路名额（包括对应特色区域）。

[Version v0.1.5]
Change the Work ethic belief back to the orignal one (based on each follower), and change to grant 2% production each follower.
把职业道德信条改回原版的形式，且每个信徒增加2%的生产力。

[Version v0.1.5.1]
Change the minimum turns for each game era from 40 turns to 30 turns.
将每个时代的回合数的下限由40t调整为30t。

fix the bug exist in [Version v0.1.4.2]
修复了[Version v0.1.4.2]中导致的bug。

[Version v0.1.5.2]
Decrease the movement lock turns of SCULPTURE PORTRAIT LANDSCAPE RELIGIOUS great works from 10 to 5.
将雕塑、肖像、风景画、宗教巨作的移动锁定时间由10回合降低为5回合。

Great works yields and tourism increased (2 -> 3 for writings, 3 -> 4 for arts & artifacts, 4 -> 6 for musics).
巨作的文化和旅游业绩调整，著作3文化3业绩，艺术作品和考古 4文化和4业绩，音乐巨作6文化6业绩。

Encampment district grant the city 20% production boost towards producing units. (Ikanda 30%)
军营为城市提供20%锤单位加成。（祖鲁ud伊坎达为30%）

[Version v0.1.6]
Add effects for the following buildings:
1. Granary add 1 food to wheat, rice, maize, deer and banana resources.
2. Lighthouse add 1 food to fish, crab, pearl and whale resources.

Maps/Tiny_Lakes.lua
------------------------------------------------------------------------------
--	FILE:	 Tiny_Lakes.lua (Forked from Lakes.lua)
--	AUTHOR:  
--	PURPOSE: Base game script - Produces widely varied islands.
------------------------------------------------------------------------------
--	Copyright (c) 2017 Firaxis Games, Inc. All rights reserved.
------------------------------------------------------------------------------

include "MapEnums"
include "MapUtilities"
include "MountainsCliffs"
include "RiversLakes"
include "FeatureGenerator"
include "TerrainGenerator"
include "NaturalWonderGenerator"
include "ResourceGenerator"
include "CoastalLowlands"
include "AssignStartingPlots"

local g_iW, g_iH;
local g_iFlags = {};
local g_continentsFrac = nil;
local islands = {};
local featureGen = nil;
local world_age_old = 2;
local world_age_normal = 3;
local world_age_new = 5;

-------------------------------------------------------------------------------
function GenerateMap()
	print("Generating Lakes Map");
	local pPlot;

	-- Set globals
	g_iW, g_iH = Map.GetGridSize();
	g_iFlags = TerrainBuilder.GetFractalFlags();
	local temperature = MapConfiguration.GetValue("temperature"); -- Default setting is Temperate.
	if temperature == 4 then
		temperature  =  1 + TerrainBuilder.GetRandomNumber(3, "Random Temperature- Lua");
	end
	
	--	local world_age
	local world_age = MapConfiguration.GetValue("world_age");
	if (world_age == 1) then
		world_age = world_age_new;
	elseif (world_age == 2) then
		world_age = world_age_normal;
	elseif (world_age == 3) then
		world_age = world_age_old;
	else
		world_age = 1 + TerrainBuilder.GetRandomNumber(3, "Random World Age - Lua");
	end

	plotTypes = GeneratePlotTypes(world_age);
	-- terrainTypes = GenerateTerrainTypes(plotTypes, g_iW, g_iH, g_iFlags, false, temperature);
	terrainTypes = GenerateTerrainTypes(plotTypes, g_iW, g_iH, g_iFlags, true, temperature); -- Remove Coastal Mountains.
	ApplyBaseTerrain(plotTypes, terrainTypes, g_iW, g_iH);
	
	AreaBuilder.Recalculate();
	TerrainBuilder.AnalyzeChokepoints();
	TerrainBuilder.StampContinents();

	local iContinentBoundaryPlots = GetContinentBoundaryPlotCount(g_iW, g_iH);
	local biggest_area = Areas.FindBiggestArea(false);
	print("After Adding Hills: ", biggest_area:GetPlotCount());
	AddTerrainFromContinents(plotTypes, terrainTypes, world_age, g_iW, g_iH, iContinentBoundaryPlots, true); -- Remove Coastal Mountains.

	AreaBuilder.Recalculate();
	TerrainBuilder.AnalyzeChokepoints();

	print ("Num Mountains: " .. tostring(GetMountainCount(g_iW, g_iH)));

	-- River generation is affected by plot types, originating from highlands and preferring to traverse lowlands.
	AddRivers();
	
	-- Lakes would interfere with rivers, causing them to stop and not reach the ocean, if placed any sooner.
	local numLargeLakes = math.ceil(GameInfo.Maps[Map.GetMapSize()].Continents * 4);
	AddLakes(numLargeLakes);

	AddFeatures();
	
	print("Adding cliffs");
	AddCliffs(plotTypes, terrainTypes);
	
	local args = {
		numberToPlace = GameInfo.Maps[Map.GetMapSize()].NumNaturalWonders,
	};
	local nwGen = NaturalWonderGenerator.Create(args);

	AddFeaturesFromContinents();
	MarkCoastalLowlands();
	
--	for i = 0, (g_iW * g_iH) - 1, 1 do
--		pPlot = Map.GetPlotByIndex(i);
--		print ("i: plotType, terrainType, featureType: " .. tostring(i) .. ": " .. tostring(plotTypes[i]) .. ", " .. tostring(terrainTypes[i]) .. ", " .. tostring(pPlot:GetFeatureType(i)));
--	end

	local resourcesConfig = MapConfiguration.GetValue("resources");
	local startConfig = MapConfiguration.GetValue("start");-- Get the start config
	local args = {
		resources = resourcesConfig,
		START_CONFIG = startConfig,
	};
	local resGen = ResourceGenerator.Create(args);
	print("Creating start plot database.");
	-- START_MIN_Y and START_MAX_Y is the percent of the map ignored for major civs' starting positions.
	
	local args = {
		MIN_MAJOR_CIV_FERTILITY = 150,
		MIN_MINOR_CIV_FERTILITY = 50, 
		MIN_BARBARIAN_FERTILITY = 1,
		START_MIN_Y = 15,
		START_MAX_Y = 15,
		START_CONFIG = startConfig,
	};
	local start_plot_database = AssignStartingPlots.Create(args)

	local GoodyGen = AddGoodies(g_iW, g_iH);
end

-------------------------------------------------------------------------------
function GeneratePlotTypes(world_age)
	print("Generating Plot Types");
	local plotTypes = {};
	local extra_mountains = 0;
	local grain_amount = 3;
	local adjust_plates = 1.0;
	local shift_plot_types = true;
	local tectonic_islands = true;
	local hills_ridge_flags = g_iFlags;
	local peaks_ridge_flags = g_iFlags;
	local has_center_rift = false;
	local sea_level = 1;
	local world_age = 1;

	local water_percent_modifier = 0; 

	for x = 0, g_iW - 1 do
		for y = 0, g_iH - 1 do
			local index = (y * g_iW) + x + 1; -- Lua Array starts at 1
			plotTypes[index] = g_PLOT_TYPE_OCEAN;
		end
	end

	--	local sea_level
    local sea_level = MapConfiguration.GetValue("sea_level");
	if sea_level == 1 then -- Low Sea Level
		water_percent_modifier = 4
	elseif sea_level == 2 then -- Normal Sea Level
		water_percent_modifier = 0;
	elseif sea_level == 3 then -- High Sea Level
		water_percent_modifier = -4;
	else
		water_percent_modifier = TerrainBuilder.GetRandomNumber(9, "Random Sea Level - Lua") - 4;
	end

	local iWaterShift1 = math.ceil(g_iH * 15 / 100);
	local iWaterHeight1 = g_iH - iWaterShift1 * 2;

	-- local args = {};	
	-- args.iWaterPercent = 92 + water_percent_modifier;
	-- args.iRegionWidth = math.ceil(g_iW);
	-- args.iRegionHeight = math.ceil(g_iH);
	-- args.iRegionWestX = math.floor(0);
	-- args.iRegionSouthY = math.floor(0);
	-- args.iRegionGrain = 3;
	-- args.iRegionHillsGrain = 4;
	-- args.iRegionPlotFlags = g_iFlags;
	-- args.iRegionFracXExp = 6;
	-- args.iRegionFracYExp = 5;
	-- plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);
	-- islands = plotTypes;

    -- Setup for Tiny Lakes
    local args = {};
    args.iWaterPercent = 75 + water_percent_modifier;
    args.iRegionWidth = math.ceil(g_iW);
    args.iRegionHeight = math.ceil(g_iH);
    args.iRegionWestX = math.floor(0);
    args.iRegionSouthY = math.floor(0);
    args.iRegionGrain = 5;
    args.iRegionHillsGrain = 4;
    args.iRegionPlotFlags = g_iFlags;
    args.iRegionFracXExp = 6;
    args.iRegionFracYExp = 5;
    plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);
    islands = plotTypes;

    local args = {};
    args.iWaterPercent = 95 + water_percent_modifier;
    args.iRegionWidth = math.ceil(g_iW);
    args.iRegionHeight = math.ceil(g_iH);
    args.iRegionWestX = math.floor(0);
    args.iRegionSouthY = math.floor(0);
    args.iRegionGrain = 5;
    args.iRegionHillsGrain = 4;
    args.iRegionPlotFlags = g_iFlags;
    args.iRegionFracXExp = 7;
    args.iRegionFracYExp = 6;
    plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);
    islands = plotTypes;
	-- Generate Large Lakes		
	-- local args = {};	
	-- args.iWaterPercent = 81 + water_percent_modifier;
	-- args.iRegionWidth = math.ceil(g_iW);
	-- args.iRegionHeight = math.ceil(g_iH);
	-- args.iRegionWestX = math.floor(0);
	-- args.iRegionSouthY = math.floor(0);
	-- args.iRegionGrain = 3;
	-- args.iRegionHillsGrain = 4;
	-- args.iRegionPlotFlags = g_iFlags;
	-- args.iRegionFracXExp = 6;
	-- args.iRegionFracYExp = 5;
	-- plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);
	-- islands = plotTypes;

	-- -- Generate Medium Lakes	
	-- local args = {};	
	-- args.iWaterPercent = 88 + water_percent_modifier;
	-- args.iRegionWidth = math.ceil(g_iW);
	-- args.iRegionHeight = math.ceil(g_iH);
	-- args.iRegionWestX = math.floor(0);
	-- args.iRegionSouthY = math.floor(0);
	-- args.iRegionGrain = 4;
	-- args.iRegionHillsGrain = 4;
	-- args.iRegionPlotFlags = g_iFlags;
	-- args.iRegionFracXExp = 7;
	-- args.iRegionFracYExp = 6;
 --    plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);
	-- islands = plotTypes;

	-- -- Generate Tiny Lakes
	-- local args = {};	
	-- args.iWaterPercent = 95 + water_percent_modifier;
	-- args.iRegionWidth = math.ceil(g_iW);
	-- args.iRegionHeight = math.ceil(g_iH);
	-- args.iRegionWestX = math.floor(0);
	-- args.iRegionSouthY = math.floor(0);
	-- args.iRegionGrain = 5;
	-- args.iRegionHillsGrain = 4;
	-- args.iRegionPlotFlags = g_iFlags;
	-- args.iRegionFracXExp = 7;
	-- args.iRegionFracYExp = 6;
 --    plotTypes = GenerateFractalLayerWithoutHills(args, plotTypes);

	-- Land and water are set. Apply hills and mountains.
	local args = {};
	args.world_age = world_age;
	args.iW = g_iW;
	args.iH = g_iH
	args.iFlags = g_iFlags;
	args.blendRidge = 10;
	-- args.blendFract = 5; -- original setting
	args.blendFract = 1;
	-- args.extra_mountains = 5; -- original setting
	args.extra_mountains = 2;
	-- mountainRatio = 8 + world_age * 3; -- original setting
	mountainRatio = 7 + world_age * 3;
	plotTypes = ApplyTectonics(args, plotTypes);
	plotTypes = AddLonelyMountains(plotTypes, mountainRatio);

	return  plotTypes;
end
-------------------------------------------------------------------------------
function AddFeatures()
	print("Adding Features");

	-- Get Rainfall setting input by user.
	local rainfall = MapConfiguration.GetValue("rainfall");
	if rainfall == 4 then
		rainfall = 1 + TerrainBuilder.GetRandomNumber(3, "Random Rainfall - Lua");
	end
	
	local args = {rainfall = rainfall}
	featuregen = FeatureGenerator.Create(args);
	featuregen:AddFeatures(true, true);  --second parameter is whether or not rivers start inland);
end
-------------------------------------------------------------------------------
function AddFeaturesFromContinents()
	print("Adding Features from Continents");

	featuregen:AddFeaturesFromContinents();
end


-------------------------------------------------------------------------------
function GenerateFractalLayerWithoutHills (args, plotTypes)
	--[[ This function is intended to be paired with ApplyTectonics. If all the hills and
	mountains plots are going to be overwritten by the tectonics results, then why waste
	calculations generating them? ]]--
	local args = args or {};
	local plotTypes2 = {};

	-- Handle args or assign defaults.
	local iWaterPercent = args.iWaterPercent or 55;
	local iRegionWidth = args.iRegionWidth; -- Mandatory Parameter, no default
	local iRegionHeight = args.iRegionHeight; -- Mandatory Parameter, no default
	local iRegionWestX = args.iRegionWestX; -- Mandatory Parameter, no default
	local iRegionSouthY = args.iRegionSouthY; -- Mandatory Parameter, no default
	local iRegionGrain = args.iRegionGrain or 1;
	local iRegionPlotFlags = args.iRegionPlotFlags or g_iFlags;
	local iRegionTerrainFlags = g_iFlags; -- Removed from args list.
	local iRegionFracXExp = args.iRegionFracXExp or 6;
	local iRegionFracYExp = args.iRegionFracYExp or 5;
	local iRiftGrain = args.iRiftGrain or -1;
	local bShift = args.bShift or true;
	
	--print("Received Region Data");
	--print(iRegionWidth, iRegionHeight, iRegionWestX, iRegionSouthY, iRegionGrain);
	--print("- - -");
	
	--print("Filled regional table.");
	-- Loop through the region's plots
	for x = 0, iRegionWidth - 1, 1 do
		for y = 0, iRegionHeight - 1, 1 do
			local i = y * iRegionWidth + x + 1; -- Lua arrays start at 1.
			plotTypes2[i] =g_PLOT_TYPE_OCEAN;
		end
	end

	-- Init the land/water fractal
	local regionContinentsFrac;
	if(iRiftGrain > 0 and iRiftGrain < 4) then
		local riftsFrac = Fractal.Create(g_iW, g_iH, rift_grain, {}, iRegionFracXExp, iRegionFracYExp);
		regionContinentsFrac = Fractal.CreateRifts(g_iW, g_iH, iRegionGrain, iRegionPlotFlags, riftsFrac, iRegionFracXExp, iRegionFracYExp);
	else
		regionContinentsFrac = Fractal.Create(g_iW, g_iH, iRegionGrain, iRegionPlotFlags, iRegionFracXExp, iRegionFracYExp);	
	end
	--print("Initialized main fractal");
	local iWaterThreshold = regionContinentsFrac:GetHeight(iWaterPercent);

	-- Loop through the region's plots
	for x = 0, iRegionWidth - 1, 1 do
		for y = 0, iRegionHeight - 1, 1 do
			local i = y * iRegionWidth + x + 1; -- Lua arrays start at 1.
			local val = regionContinentsFrac:GetHeight(x,y);
			if val >= iWaterThreshold or Adjacent(i) == true then
				--do nothing
			else
				plotTypes2[i] = g_PLOT_TYPE_LAND;
			end
		end
	end

	if bShift then -- Shift plots to obtain a more natural shape.
		ShiftPlotTypes(plotTypes);
	end

	print("Shifted Plots - Width: ", iRegionWidth, "Height: ", iRegionHeight);

	-- Apply the region's plots to the global plot array.
	for x = 0, iRegionWidth - 1, 1 do
		local wholeworldX = x + iRegionWestX;
		for y = 0, iRegionHeight - 1, 1 do
			local index = y * iRegionWidth + x + 1
			if plotTypes2[index] ~= g_PLOT_TYPE_OCEAN then
				local wholeworldY = y + iRegionSouthY;
				local i = wholeworldY * g_iW + wholeworldX + 1
				plotTypes[i] = plotTypes2[index];
			end
		end
	end
	--print("Generated Plot Types");

	return plotTypes;
end

-------------------------------------------------------------------------------------------
function Adjacent(index)
	aIslands = islands;
	index = index -1;

	if(aIslands == nil) then
		return false;
	end
	
	if(index < 0) then
		return false
	end

	local plot = Map.GetPlotByIndex(index);
	if(aIslands[index] ~= nil and aIslands[index] == g_PLOT_TYPE_LAND) then
		return true;
	end

	for direction = 0, DirectionTypes.NUM_DIRECTION_TYPES - 1, 1 do
		local adjacentPlot = Map.GetAdjacentPlot(plot:GetX(), plot:GetY(), direction);
		if(adjacentPlot ~= nil) then
			local newIndex = adjacentPlot:GetIndex();
			if(aIslands  ~= nil and aIslands[newIndex] == g_PLOT_TYPE_LAND) then
				return true;
			end
		end
	end

	return false;
end

-------------------------------------------------------------------------------------------
function AddLakes(largeLakes)

	print("Map Generation - Adding Lakes");
	largeLakes = largeLakes or 0;

	local numLakesAdded = 0;
	local numLargeLakesAdded = 0;

	local lakePlotRand =  math.floor(GlobalParameters.LAKE_PLOT_RANDOM or 25 / 4);
	local iW, iH = Map.GetGridSize();

	for i = 0, (iW * iH) - 1, 1 do
		plot = Map.GetPlotByIndex(i);
		if(plot) then
			if (plot:IsWater() == false) then
				if (plot:IsCoastalLand() == false) then
					if (plot:IsRiver() == false) then
						if (AdjacentToNaturalWonder(plot) == false) then
							local r = TerrainBuilder.GetRandomNumber(lakePlotRand, "MapGenerator AddLakes");
							if r == 0 then
								numLakesAdded = numLakesAdded + 1;
								if(largeLakes > numLargeLakesAdded) then
									local bLakes = AddMoreLake(plot);
									if(bLakes == true) then
										numLargeLakesAdded = numLargeLakesAdded + 1;
									end
								end

								TerrainBuilder.SetTerrainType(plot, g_TERRAIN_TYPE_COAST);
							end
						end
					end
				end
			end
		end
	end
	
	-- this is a minimalist update because lakes have been added
	if numLakesAdded > 0 then
		print(tostring(numLakesAdded).." lakes added")
		AreaBuilder.Recalculate();
	end
end